File: src/main/java/net/jgp/books/spark/ch11/lab100_simple_select/SimpleSelectApp.java
package net.jgp.books.spark.ch11.lab100_simple_select;

import org.apache.spark.sql.Dataset;
import org.apache.spark.sql.Row;
import org.apache.spark.sql.SparkSession;
import org.apache.spark.sql.types.DataTypes;
import org.apache.spark.sql.types.StructField;
import org.apache.spark.sql.types.StructType;

/**
 * Simple SQL select on ingested data
 * 
 * @author jgp
 */
public class SimpleSelectApp {

  /**
   * main() is your entry point to the application.
   * 
   * @param args
   */
  public static void main(String[] args) {
    SimpleSelectApp app = new SimpleSelectApp();
    app.start();
  }

  /**
   * The processing code.
   */
  private void start() {
    // Creates a session on a local master
    SparkSession spark = SparkSession.builder()
        .appName("Simple SELECT using SQL")
        .master("local")
        .getOrCreate();

    StructType schema = DataTypes.createStructType(new StructField[] {
        DataTypes.createStructField(
            "geo",
            DataTypes.StringType,
            true),
        DataTypes.createStructField(
            "yr1980",
            DataTypes.DoubleType,
            false) });

    // Reads a CSV file with header, called books.csv, stores it in a
    // dataframe
    Dataset<Row> df = spark.read().format("csv")
        .option("header", true)
        .schema(schema)
        .load("data/populationbycountry19802010millions.csv");
    df.createOrReplaceTempView("geodata");
    df.printSchema();

    Dataset<Row> smallCountries =
        spark.sql(
            "SELECT * FROM geodata WHERE yr1980 < 1 ORDER BY 2 LIMIT 5");

    // Shows at most 10 rows from the dataframe (which is limited to 5
    // anyway)
    smallCountries.show(10, false);
  }
}

File: src/main/python/lab100_simple_select/simpleSelectApp.py
"""
Simple SQL select on ingested data
@author rambabu.posa
"""
import os
from pyspark.sql import SparkSession
from pyspark.sql.types import (StructType,StructField,
                               StringType,DoubleType)

def get_absolute_file_path(path, filename):
    current_dir = os.path.dirname(__file__)
    relative_path = f"{path}{filename}"
    absolute_file_path = os.path.join(current_dir, relative_path)
    return absolute_file_path

def main(spark):
    path="../../../../data/"
    filename="populationbycountry19802010millions.csv"
    absolute_file_path = get_absolute_file_path(path, filename)

    schema = StructType([
        StructField('geo',StringType(), True),
        StructField('yr1980', DoubleType(), False)
      ])

    # Reads a CSV file with header, called books.csv, stores it in a dataframe
    df = spark.read.csv(header=True, inferSchema=True, schema=schema, path=absolute_file_path)

    df.createOrReplaceTempView('geodata')
    df.printSchema()

    query = """
      SELECT * FROM geodata
      WHERE yr1980 < 1
      ORDER BY 2
      LIMIT 5
    """

    smallCountries = spark.sql(query)

    # Shows at most 10 rows from the dataframe (which is limited to 5
    # anyway)
    smallCountries.show(10, False)


if __name__ == "__main__":

    # Creates a session on a local master
    spark = SparkSession.builder.appName("Simple SELECT using SQL") \
        .master("local[*]") \
        .getOrCreate()

    main(spark)

    # Good to stop SparkSession at the end of the application
    spark.stop()